File: app/src/main/java/com/example/nutritionapp/Register/FragmentEmail.java
package com.example.nutritionapp.Register;


import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.text.Editable;
import android.text.TextWatcher;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.example.nutritionapp.Activity.Login;
import com.example.nutritionapp.R;
import com.example.nutritionapp.databinding.FragmentEmailBinding;

import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;

import java.util.Objects;

public class FragmentEmail extends Fragment {
    private FragmentEmailBinding binding;
    RegisterMain registerMain;
    EditText textEmail, textMobile,textName;
    ExtendedFloatingActionButton next,back;
    TextInputLayout nameL,emailL,mobileL;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentEmailBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        textEmail= root.findViewById(R.id.text_email);
        textMobile= root.findViewById(R.id.text_mobile);
        textName=root.findViewById(R.id.text_name);
        next= root.findViewById(R.id.email_next);
        back=root.findViewById(R.id.email_back);
        nameL=root.findViewById(R.id.nameLayout);
        emailL=root.findViewById(R.id.emailLayout);
        mobileL=root.findViewById(R.id.mobileLayout);

        // to get data and set the texts when came from passwordFragment
        String passwd = "", active = "light";
        boolean selectedMale = false;
        boolean selectedFemale = false;
        int weight =50, age = 18, height = 180;
        Bundle bundle = this.getArguments();
        if(bundle!=null) {
            String name = bundle.getString("name", "");
            String email = bundle.getString("email", "");
            String mobileNo = bundle.getString("mobile", "");
            passwd = bundle.getString("pass", "");
            selectedMale = bundle.getBoolean("selectedMale", false);
            selectedFemale = bundle.getBoolean("selectedFemale", false);
            weight = bundle.getInt("weight", 50);
            age = bundle.getInt("age", 18);
            height = bundle.getInt("height", 180);
            active = bundle.getString("active", "light");
            textName.setText(name);
            textEmail.setText(email);
            textMobile.setText(mobileNo);
        }
        registerMain = (RegisterMain) getActivity();

        String finalPasswd = passwd;
        boolean finalSelectedFemale = selectedFemale;
        boolean finalSelectedMale = selectedMale;
        int finalWeight = weight;
        int finalAge = age;
        int finalHeight = height;
        String finalActive = active;
        next.setOnClickListener(view -> {
            if(textEmail.getText().toString()!=null & textMobile.getText().toString()!= null) {
                registerMain.myEdit.putString("email", textEmail.getText().toString());
                registerMain.myEdit.putString("mobile", textMobile.getText().toString());
                registerMain.myEdit.commit();
            }

            boolean valid = true;
            if ( !validateEmail()) {
                valid = false;
            }

            if (textMobile.length() == 0||textMobile.length()<10||textMobile.length()>10) {
                mobileL.setError("Please Enter a Valid Mobile Number");
                textMobile.setError(null);
                valid = false;

            }
            if(textName.length()==0)
            {
                nameL.setError("Name is required");
                textName.setError(null);
                valid=false;
            }
            if(valid) {
                FragmentPassword fragmentPassword = new FragmentPassword();
                sendInfo(fragmentPassword, finalPasswd, finalSelectedMale, finalSelectedFemale, finalAge, finalWeight, finalHeight, finalActive);

                getParentFragmentManager().beginTransaction()
                .replace(R.id.register_container, fragmentPassword)
                .commit();
                registerMain.myEdit.putString("name", textName.getText().toString());
                registerMain.myEdit.commit();
            }

        });
        back.setOnClickListener(view -> {

            Intent intent = new Intent(getActivity(), Login.class);
            startActivity(intent);
        });
        textName.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                nameL.setError(null);
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
        textMobile.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                mobileL.setError(null);
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
        textEmail.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                emailL.setError(null);
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });

        return root;

    }
    private boolean validateEmail() {
        String email = textEmail.getText().toString().trim();
        FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
        ;
        //check email already exist or not.
        if (!email.isEmpty()){
            firebaseAuth.fetchSignInMethodsForEmail(email)
                    .addOnCompleteListener(task -> {
                        boolean isNewUser = Objects.requireNonNull(task.getResult().getSignInMethods()).isEmpty();

                        if (isNewUser) {
                            emailL.setError(null);
                        } else {
                            emailL.setError("User already exists");
                        }
                    });
            return false;
    }
        if (email.isEmpty()) {
            emailL.setError("Email is required");
            textEmail.setError(null);
            return false;
        } else if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            emailL.setError("Please Enter a Valid Email Address");
            textEmail.setError(null);
            return false;
        } else {
            emailL.setError(null);
            return true;
        }
    }

    private void sendInfo(Fragment fragment, String passwd, boolean selectedMale, boolean selectedFemale, int age, int weight, int height, String active){

        // get data from SharedPreferences if next/back is pressed
        /** data  only gets stored in shared preferences if next is clicked
         * other wise take data from editText
         * */
        // add data to bundle
        Bundle bundle = new Bundle();
        bundle.putString("pass", passwd);

        bundle.putBoolean("selectedMale", selectedMale);
        bundle.putBoolean("selectedFemale", selectedFemale);
        bundle.putInt("age", age);
        bundle.putInt("weight", weight);
        bundle.putInt("height", height);
        bundle.putString("active", active);
        fragment.setArguments(bundle);
    }
}

File: app/src/main/java/com/example/nutritionapp/Register/FragmentEmail.kt
package com.example.nutritionapp.Register

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import com.example.nutritionapp.Activity.LoginActivity
import com.example.nutritionapp.R
import com.example.nutritionapp.databinding.FragmentEmailBinding
import com.google.android.gms.tasks.Task
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.SignInMethodQueryResult

class FragmentEmail : Fragment() {
    private var binding: FragmentEmailBinding? = null
    var registerMain: RegisterMain? = null
    lateinit var textEmail: EditText
    lateinit var textMobile: EditText
    lateinit var textName: EditText
    lateinit var next: ExtendedFloatingActionButton
    lateinit var back: ExtendedFloatingActionButton
    lateinit var nameL: TextInputLayout
    lateinit var emailL: TextInputLayout
    lateinit var mobileL: TextInputLayout
    var progressBar: ProgressBar? = null
    val EMAIL_PATTERN = "^[_A-Za-z0-9-]+(\\.[_A-Za-z0-9-]+)*@[A-Za-z0-9]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$"
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentEmailBinding.inflate(inflater, container, false)
        val root: View = binding!!.root
        textEmail = root.findViewById(R.id.text_email)
        textMobile = root.findViewById(R.id.text_mobile)
        textName = root.findViewById(R.id.text_name)
        next = root.findViewById(R.id.email_next)
        back = root.findViewById(R.id.email_back)
        nameL = root.findViewById(R.id.nameLayout)
        emailL = root.findViewById(R.id.emailLayout)
        mobileL = root.findViewById(R.id.mobileLayout)
        progressBar = root.findViewById(R.id.next_progress_bar)
        // to get data and set the texts when came from passwordFragment
        var passwd = ""
        var active = "light"
        var selectedMale = false
        var selectedFemale = false
        var weight = 50
        var age = 18
        var height = 180
        val bundle = this.arguments
        if (bundle != null) {
            val name = bundle.getString("name", "")
            val email = bundle.getString("email", "")
            val mobileNo = bundle.getString("mobile", "")
            passwd = bundle.getString("pass", "")
            selectedMale = bundle.getBoolean("selectedMale", false)
            selectedFemale = bundle.getBoolean("selectedFemale", false)
            weight = bundle.getInt("weight", 50)
            age = bundle.getInt("age", 18)
            height = bundle.getInt("height", 180)
            active = bundle.getString("active", "light")
            textName.setText(name)
            textEmail.setText(email)
            textMobile.setText(mobileNo)
        }
        registerMain = activity as RegisterMain?
        val finalPasswd = passwd
        val finalSelectedFemale = selectedFemale
        val finalSelectedMale = selectedMale
        val finalWeight = weight
        val finalAge = age
        val finalHeight = height
        val finalActive = active
        next.setOnClickListener(View.OnClickListener { view: View? ->
//            if (textEmail.getText().toString()!=null!! and textMobile.getText().toString() != null) {
            registerMain!!.myEdit?.putString("email", textEmail.getText().toString())
            registerMain!!.myEdit?.putString("mobile", textMobile.getText().toString())
            registerMain!!.myEdit?.commit()
//            }
            var valid = true
            if (!validateEmail()) {
                valid = false
            }
            if (textMobile.length() == 0 || textMobile.length() < 10 || textMobile.length() > 10) {
                mobileL.setError("Please Enter a Valid Mobile Number")
                textMobile.setError(null)
                valid = false
            }
            if (textName.length() == 0) {
                nameL.setError("Name is required")
                textName.setError(null)
                valid = false
            }
            if (valid) {
                val fragmentPassword = FragmentPassword()
                sendInfo(fragmentPassword, finalPasswd, finalSelectedMale, finalSelectedFemale, finalAge, finalWeight, finalHeight, finalActive)
                parentFragmentManager.beginTransaction()
                        .replace(R.id.register_container, fragmentPassword)
                        .commit()
                registerMain!!.myEdit?.putString("name", textName.getText().toString())
                registerMain!!.myEdit?.commit()
            }
        })
        back.setOnClickListener(View.OnClickListener { view: View? ->
            val intent = Intent(activity, LoginActivity::class.java)
            startActivity(intent)
        })
        textName.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                nameL.setError(null)
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        textMobile.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                mobileL.setError(null)
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        textEmail.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                emailL.setError(null)
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        return root
    }

    private fun validateEmail(): Boolean {
        val email = textEmail!!.text.toString().trim { it <= ' ' }
        val firebaseAuth = FirebaseAuth.getInstance()
        if (email.isEmpty()) {
            emailL.error = "Please Enter an Email-ID"
            textEmail.requestFocus()
            return false
        }
         return if (email.matches(EMAIL_PATTERN.toRegex())) {
            //check email already exist or not.
            val isNewUser = BooleanArray(1)
            firebaseAuth.fetchSignInMethodsForEmail(email)
                    .addOnCompleteListener { task: Task<SignInMethodQueryResult> ->
                        if (task.isSuccessful) {
                            isNewUser[0] = task.result.signInMethods!!.isEmpty()
                            if (isNewUser[0]) {
                                isNewUser[0] = false
                                Log.e("TAG", "Is New User!")
                            } else {
                                Log.e("TAG", "Is Old User!")
                                emailL!!.error = "User already exists"
                                textEmail!!.requestFocus()
                            }
                        }
                    }
             !isNewUser[0]
        } else {
            emailL!!.error = "Please Enter a valid Email-ID"
            textEmail!!.requestFocus()
            false
        }
    }

    private fun sendInfo(fragment: Fragment, passwd: String, selectedMale: Boolean, selectedFemale: Boolean, age: Int, weight: Int, height: Int, active: String) {

        // get data from SharedPreferences if next/back is pressed
        /** data  only gets stored in shared preferences if next is clicked
         * other wise take data from editText
         */
        // add data to bundle
        val bundle = Bundle()
        bundle.putString("pass", passwd)
        bundle.putBoolean("selectedMale", selectedMale)
        bundle.putBoolean("selectedFemale", selectedFemale)
        bundle.putInt("age", age)
        bundle.putInt("weight", weight)
        bundle.putInt("height", height)
        bundle.putString("active", active)
        fragment.arguments = bundle
    }
}